dandiapi/api/services/garbage_collection/__init__.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
+
+from datetime import timedelta
+import json
+from uuid import uuid4
+
+from celery.utils.log import get_task_logger
+from django.core import serializers
+from django.db import transaction
+from django.utils import timezone
+from more_itertools import chunked
+
+from dandiapi.api.models import AssetBlob, GarbageCollectionEvent, Upload
+from dandiapi.api.storage import DandiMultipartMixin
+
+logger = get_task_logger(__name__)
+
+UPLOAD_EXPIRATION_TIME = DandiMultipartMixin._url_expiration  # noqa: SLF001
+ASSET_BLOB_EXPIRATION_TIME = timedelta(days=7)
+
+GARBAGE_COLLECTION_EVENT_CHUNK_SIZE = 1000
+
+# How long to keep GarbageCollectionEvent records around for after the garbage collection
+# is performed to allow for restoration of deleted records.
+# This should be equal to the time set in the "trailing delete" lifecycle rule.
+RESTORATION_WINDOW = timedelta(
+    days=30
+)  # TODO: pick this up from env var set by Terraform to ensure consistency?
+
+
+def _garbage_collect_uploads() -> int:
+    qs = Upload.objects.filter(
+        created__lt=timezone.now() - UPLOAD_EXPIRATION_TIME,
+    )
+
+    # Chunk the queryset to avoid creating a single
+    # GarbageCollectionEvent with a huge JSONField
```

Comment:
This may also be something that's easier with an `Event <- Record` model, as I mentioned in my other comment. You could have many records pointing to one event, which would allow for just one event record to be stored per "actual" event. In the current case, there could be 10 `GarbageCollectionEvent` rows for what's really one "event', i.e. one invocation of `_garbage_collect_uploads` or `_garbage_collect_asset_blobs`.

Reply:
This would also allow you to point both asset blobs and uploads back to the same single event record, rather than two separate records.